gateway/handler.go

```diff
@@ -314,6 +322,68 @@ func (i *handler) addUserHeaders(w http.ResponseWriter) {
 	}
 }
 
+func (i *handler) isTrustedMode(r *http.Request) bool {
+	// Get the host, by default the request's Host. If this request went through
+	// WithHostname, also check for the key in the context. If that is not present,
+	// also check X-Forwarded-Host to support reverse proxies.
+	host := r.Host
+	if h, ok := r.Context().Value(GatewayHostnameKey).(string); ok {
+		host = h
+	} else if xHost := r.Header.Get("X-Forwarded-Host"); xHost != "" {
+		host = xHost
+	}
+
+	// If the gateway is defined, return whatever is set.
+	if gw, ok := i.config.PublicGateways[host]; ok {
+		return gw.TrustedMode
+	}
+
+	// Cleanup the host and check if it's a local gateway. If so, it's trusted.
+	host = stripPort(host)
+	if host == "127.0.0.1" || host == "::1" || host == "localhost" {
+		return true
+	}
+
+	// Otherwise, the default.
+	return i.config.TrustedMode
+}
+
+func (i *handler) isValidTrustlessRequest(contentPath ipath.Path, responseFormat string) bool {
+	// Only allow "/{#1}/{#2}"-like paths.
+	trimmedPath := strings.Trim(contentPath.String(), "/")
+	pathComponents := strings.Split(trimmedPath, "/")
+	if len(pathComponents) != 2 {
+		return false
+	}
+
+	if contentPath.Namespace() == "ipns" {
+		// Only ipns records allowed.
+		if responseFormat != "application/vnd.ipfs.ipns-record" {
+			return false
+		}
+
+		// Only valid peers, no DNSLink.
+		if _, err := peer.Decode(pathComponents[1]); err != nil {
```

Comment:
:thought_balloon: now that we've moved go-ipns into `boxo/ipns`, it may be a good idea to add `ipns.DecodeName` and `ipns.EncodeName` there so we don't use PeerIDs in code related to IPNS.

(Internally it would still use `peer.Decode`, but allows us to control things like base and supported legacy encodings independently of go-libp2p/peer defaults.

Reply:
Could potentially also be used for #236.